File: framework/languages/dotnet/src/Zlink.Framework/Contracts/Registry/IZLinkRegistryQuery.cs
namespace Zlink.Framework.Contracts.Registry;

public interface IZLinkRegistryQuery
{
    ValueTask<ZLinkRegistryStatus> StatusAsync(
        CancellationToken cancellationToken = default);

    ValueTask<ZLinkRegistryServiceSummaryEntry[]> ServiceSummaryAsync(
        ZLinkRegistryServiceSummaryFilter? filter = null,
        CancellationToken cancellationToken = default);

    ValueTask<ZLinkRegistryTopologyEntry[]> TopologyAsync(
        ZLinkRegistryTopologyFilter? filter = null,
        CancellationToken cancellationToken = default);

    ValueTask<ZLinkMemberPeerEntry[]> MemberPeersAsync(
        string channelName,
        CancellationToken cancellationToken = default);
}

public interface IZLinkRegistryQueryClient
{
    ValueTask<ZLinkRegistryTopologyEntry[]> TopologyAsync(
        ZLinkRegistryTopologyFilter? filter = null,
        CancellationToken cancellationToken = default);
}

public interface IZLinkRegistryQueryClientOptions
{
    string Endpoint { get; set; }
}

File: framework/languages/node/packages/framework/src/contracts/Registry/IZLinkRegistryQuery.ts
import type {
  ZLinkMemberPeerEntry,
  ZLinkRegistryServiceSummaryEntry,
  ZLinkRegistryServiceSummaryFilter,
  ZLinkRegistryStatus,
  ZLinkRegistryTopologyEntry,
  ZLinkRegistryTopologyFilter
} from './Models';

export interface ZLinkRegistryQuery {
  status(signal?: AbortSignal): Promise<ZLinkRegistryStatus>;
  serviceSummary(
    filter?: ZLinkRegistryServiceSummaryFilter,
    signal?: AbortSignal
  ): Promise<readonly ZLinkRegistryServiceSummaryEntry[]>;
  topology(
    filter?: ZLinkRegistryTopologyFilter,
    signal?: AbortSignal
  ): Promise<readonly ZLinkRegistryTopologyEntry[]>;
  memberPeers(channelName: string, signal?: AbortSignal): Promise<readonly ZLinkMemberPeerEntry[]>;
}

export interface ZLinkRegistryQueryClient {
  topology(
    filter?: ZLinkRegistryTopologyFilter,
    signal?: AbortSignal
  ): Promise<readonly ZLinkRegistryTopologyEntry[]>;
}

export interface ZLinkRegistryQueryClientOptions {
  endpoint: string;
}
